### PR TITLE
refactor: Compress quotient table

### PR DESCRIPTION
**Pull request by aszepieniec**
Refactor how the quotient combination codeword is computed. Now it is done without ever constructing the entire quotient table, meaning that a lot less memory is used. Additionally, the part of the sum across constraint values of the same category now preceeds the multiplication by the zerofier-inverse, leading to less multiplications and better proving time.

### CODE REVIEW

File: triton-vm/src/stark.rs [hunk 1]
```diff
@@ -176,41 +176,37 @@ impl Stark {
...
         proof_stream.enqueue(ProofItem::MerkleRoot(ext_merkle_tree.root()));
+
+        // Get the weights with which to compress the many quotients into one.
+        let quotient_combination_weights =
+            proof_stream.sample_scalars(MasterExtTable::NUM_CONSTRAINTS);
         prof_stop!(maybe_profiler, "Fiat-Shamir");
         prof_stop!(maybe_profiler, "ext tables");
 
+        // low-degree extend the trace codewords so that the all quotient codewords
+        // obtained by element-wise evaluation of the AIR constraints
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:187]: Typo + missing words?

File: triton-vm/src/table/master_table.rs [hunk 2]
```diff
@@ -1017,26 +1018,27 @@ pub fn terminal_quotient_zerofier_inverse(
     BFieldElement::batch_inversion(zerofier_codeword).into()
 }
 
-/// Computes an array containing all quotients – the Master Quotient Table. Each column corresponds
-/// to a different quotient. The quotients are ordered by category – initial, consistency,
+/// Computes the quotient codeword, which the randomized linear combination of all individual
+/// quotients.
```
Comment (jan-ferdinand) [inline triton-vm/src/table/master_table.rs:1020]: Modulo max line width:
```suggestion
/// Computes the quotient codeword, which is the randomized linear combination of all individual
/// quotients.
```

File: triton-vm/src/table/master_table.rs [hunk 3]
```diff
@@ -1046,59 +1048,117 @@ pub fn all_quotients(
...
-        &mut quotient_table.slice_mut(s![.., tran_section_end..term_section_end]),
-        terminal_quotient_zerofier_inverse(trace_domain, quotient_domain).view(),
-        challenges,
+    let quotient_codeword = (0..quotient_domain.length)
+        .into_par_iter()
+        .map(|row| {
+            // get views of relevant rows
+            let unit_distance = quotient_domain.length / trace_domain.length;
+            let domain_length_bit_mask = quotient_domain.length - 1;
+            let next_row_index = (row + unit_distance) & domain_length_bit_mask;
```
Comment (jan-ferdinand) [inline triton-vm/src/table/master_table.rs:1073]: I think we can afford a modulo operation.
```suggestion
            let next_row_index = (row + unit_distance) % quotient_domain.length;
```
Knowing that the snippets are equivalent is not necessarily obvious. I'm assuming that the applied optimiziation is not measurable.

**Comment by aszepieniec** [comment]
Benchmarking against master using `prove_halt` on my  laptop gives:
 - computing the quotient codeword goes from 12.35 ms to 8.91 ms (39% improvement)
 - overall proving time goes from 71.34 ms to 66.14 ms (7.8% improvement)
 - ram usage* goes from 76944 kB to 40472 kB (47% reduction)

*ram usage measured using `/bin/usr/time -v <cmd>` and looking at "Maximum resident set size".

File: triton-vm/src/table/master_table.rs [hunk 4]
```diff
@@ -1046,59 +1048,117 @@ pub fn all_quotients(
...
+            let current_row_main = quotient_domain_master_base_table.slice(s![row, ..]);
```
Comment (jan-ferdinand) [inline triton-vm/src/table/master_table.rs:1074]: It's probably more readable to use `.row()` here and in the following lines:
```suggestion
            let current_row_main = quotient_domain_master_base_table.row(row);
```

File: triton-vm/src/table/master_table.rs [hunk 5]
```diff
@@ -1046,59 +1048,117 @@ pub fn all_quotients(
...
+                current_row_aux,
+                challenges,
+            );
+            let initial_inner_product = initial_constraint_values
+                .into_iter()
+                .zip_eq(
+                    quotient_weights
+                        .iter()
+                        .take(MasterExtTable::NUM_INITIAL_CONSTRAINTS),
+                )
```
Comment (jan-ferdinand) [inline triton-vm/src/table/master_table.rs:1091]:
```suggestion
                .zip_eq(quotient_weights[..init_section_end].iter())
```

File: triton-vm/src/table/master_table.rs [hunk 6]
```diff
@@ -1046,59 +1048,117 @@ pub fn all_quotients(
...
+                challenges,
+            );
+            let transition_inner_product = transition_constraint_values
+                .into_iter()
+                .zip_eq(
+                    quotient_weights
+                        .iter()
+                        .skip(cons_section_end)
+                        .take(MasterExtTable::NUM_TRANSITION_CONSTRAINTS),
+                )
```
Comment (jan-ferdinand) [inline triton-vm/src/table/master_table.rs:1129]:
```suggestion
                .zip_eq(quotient_weights[cons_section_end..tran_section_end].iter())
```

File: triton-vm/src/table/master_table.rs (hunk 5 above)
Comment (jan-ferdinand) [reply triton-vm/src/table/master_table.rs:1091]: Same for the other three; see the conversations marked as “resolved” for copy-pasteable code.

File: triton-vm/src/table/master_table.rs [hunk 7]
```diff
@@ -1017,26 +1018,27 @@ pub fn terminal_quotient_zerofier_inverse(
...
 /// is strictly speaking not a table.
 /// The order of the quotients is not actually important. However, it must be consistent between
 /// [prover] and [verifier].
 ///
-/// The returned array is in row-major order.
-///
 /// [prover]: crate::stark::Stark::prove
 /// [verifier]: crate::stark::Stark::verify
-pub fn all_quotients(
+pub fn all_quotients_combined(
```
[the remark this answers is not in the document]
Comment (aszepieniec) [reply triton-vm/src/table/master_table.rs:1031]: With an eye to future changes, after which no single quotient codeword exists, I think I prefer the abstract function name whose accuracy does not stand to change.

**Pull request merged**